File: html/app/models/payment.py
from sqlalchemy import Column, Integer, Float, Date, String, ForeignKey, Enum as SQLEnum, Index, DateTime
from sqlalchemy.orm import relationship
from enum import Enum
from .base import Base

class PaymentMethod(str, Enum):
    CASH = "cash"
    CARD = "card"
    BANK_TRANSFER = "bank_transfer"

class Payment(Base):
    __tablename__ = "payments"
    
    # Performance indexes for frequently queried columns
    __table_args__ = (
        Index('ix_payment_student_period', 'student_id', 'payment_period'),
        Index('ix_payment_status_date', 'status', 'payment_date'),
        Index('ix_payment_period', 'payment_period'),
        Index('ix_payment_deleted_at', 'deleted_at'),
    )

    id = Column(Integer, primary_key=True, index=True)
    student_id = Column(Integer, ForeignKey("students.id", ondelete="CASCADE"))
    amount = Column(Float)
    payment_date = Column(Date)
    payment_period = Column(Date)  # For which month the payment is made (e.g., 01.09.2024)
    method = Column(String, default="cash")
    
    # New fields
    status = Column(String, default="completed")  # completed, pending, cancelled
    description = Column(String, nullable=True)
    reference_id = Column(String, nullable=True)  # Transaction ID, Invoice #
    
    # Soft delete fields
    deleted_at = Column(DateTime, nullable=True)
    deletion_reason = Column(String(255), nullable=True)
    deleted_by_id = Column(Integer, nullable=True)
    
    # Metadata for restoration
    last_student_name = Column(String(255), nullable=True)  # Store student name before deletion

    # Relationships
    student = relationship("Student", back_populates="payments")
    invoice_items = relationship("InvoiceItem", back_populates="payment", cascade="all, delete-orphan")

    @property
    def student_name(self):
        if self.student:
            return f"{self.student.first_name} {self.student.last_name}"
        return self.last_student_name

File: html/football_academy_app/lib/models/payment.dart
class Payment {
  final int id;
  final int studentId;
  final String? studentName;
  final double amount;
  final String paymentDate;
  final String paymentPeriod;
  final String method;
  final String status;           // completed, pending, cancelled
  final String? description;     // Описание

  Payment({
    required this.id,
    required this.studentId,
    this.studentName,
    required this.amount,
    required this.paymentDate,
    required this.paymentPeriod,
    required this.method,
    this.status = 'completed',
    this.description,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    return Payment(
      id: json['id'],
      studentId: json['student_id'],
      studentName: json['student_name'] ?? (json['student'] != null
          ? '${json['student']['first_name']} ${json['student']['last_name']}'
          : null),
      amount: (json['amount'] ?? 0).toDouble(),
      paymentDate: json['payment_date'] ?? '',
      paymentPeriod: json['payment_period'] ?? '',
      method: json['method'] ?? 'cash',
      status: json['status'] ?? 'completed',
      description: json['description'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'student_id': studentId,
      'amount': amount,
      'payment_date': paymentDate,
      'payment_period': paymentPeriod,
      'method': method,
      'status': status,
      'description': description,
    };
  }

  String get methodDisplayName {
    switch (method) {
      case 'cash':
        return '💵 Наличные';
      case 'card':
        return '💳 Карта';
      case 'bank_transfer':         // Changed from 'transfer' to match backend
        return '🏦 Банковский перевод';
      default:
        return method;
    }
  }
  
  String get statusDisplayName {
    switch (status) {
      case 'completed':
        return '✅ Оплачено';
      case 'pending':
        return '⏳ Ожидает';
      case 'cancelled':
        return '❌ Отменён';
      default:
        return status;
    }
  }

  String get formattedPeriod {
    try {
      final dt = DateTime.parse(paymentPeriod);
      const months = [
        'Январь', 'Февраль', 'Март', 'Апрель',
        'Май', 'Июнь', 'Июль', 'Август',
        'Сентябрь', 'Октябрь', 'Ноябрь', 'Декабрь'
      ];
      return '${months[dt.month - 1]} ${dt.year}';
    } catch (e) {
      return paymentPeriod;
    }
  }
  
  // Format amount with MDL currency
  String get formattedAmount {
    return '${amount.toStringAsFixed(0)} MDL';
  }
}
